File: examples/quickstart-typescript/tests/contract/test_contract.ts
import { describe, it, expect, beforeEach } from "vitest";
import { steleContext } from "./conftest.js";
import * as runtime from "./_stele_runtime.js";
import type { SteleContext } from "./_stele_runtime.js";

void runtime; // helpers: steleGetPath, steleEq, steleNeq, steleGt, steleGte, steleLt, steleLte, steleAbs, steleAdd, steleSub, steleMul, steleDiv, steleNeg, steleSum, steleCount, steleAvg, steleMin, steleMax, steleDistinct, steleUnique, steleHasLength, steleIsEmpty, steleExistsIn, steleContains, steleStartsWith, steleEndsWith, steleMatches, steleNotNull, steleBetween, steleApproxEq, steleForall, steleExists, steleWhere, steleNone, steleIsModified, steleStateBefore, steleStateAfter, steleWithin, steleBefore, steleAfter, steleRunScenario, steleCallChecker, steleMergeContexts, steleLength, steleConcat, steleSortBy, steleSortByDesc, steleMod, stelePow, steleRound, steleCeil, steleFloor, steleTrim, steleLower, steleUpper, steleSplit, steleJoin, steleJsonPath, steleDecimalEq, steleTypeOf, steleMap, steleFirst, steleLast

describe("Stele Contract", () => {
  let ctx: SteleContext;
  beforeEach(() => {
    ctx = steleContext;
  });

  it("ORDER_TOTAL_POSITIVE", () => {
    expect(runtime.steleForall(runtime.steleGetPath(ctx, ["orders"]), (order: unknown) => runtime.steleGt(runtime.steleGetPath(order, ["total"]), 0), "(gt (path order total) 0)")).toBe(true);
  });

  it("ORDER_ID_PRESENT", () => {
    expect(runtime.steleForall(runtime.steleGetPath(ctx, ["orders"]), (order: unknown) => runtime.steleNotNull(runtime.steleGetPath(order, ["id"])), "(not-null (path order id))")).toBe(true);
  });

  it("USER_STATUS_ENUM", () => {
    expect(runtime.steleEq(runtime.steleGetPath(ctx, ["user", "status"]), "active") || runtime.steleEq(runtime.steleGetPath(ctx, ["user", "status"]), "suspended") || runtime.steleEq(runtime.steleGetPath(ctx, ["user", "status"]), "deleted")).toBe(true);
  });

  it("SKU_FORMAT", () => {
    const stele_checker_result = runtime.steleCallChecker("validate-sku", ctx, {});
    expect(stele_checker_result.passed, stele_checker_result.message ?? "Checker failed: validate-sku").toBe(true);
  });

  it("EMAIL_FORMAT", () => {
    const stele_checker_result = runtime.steleCallChecker("validate-email", ctx, {});
    expect(stele_checker_result.passed, stele_checker_result.message ?? "Checker failed: validate-email").toBe(true);
  });
});

File: examples/quickstart-python/tests/contract/test_contract.py
from ._stele_runtime import stele_call_checker, stele_get_path, stele_is_modified, stele_sum


def test_ORDER_TOTAL_POSITIVE(stele_context):
    assert all(
        (stele_get_path(order, ["total"])) > (0)
        for order in stele_context["orders"]
    )


def test_ORDER_ID_PRESENT(stele_context):
    assert all(
        stele_get_path(order, ["id"]) is not None
        for order in stele_context["orders"]
    )


def test_USER_STATUS_ENUM(stele_context):
    assert ((stele_get_path(stele_context["user"], ["status"])) == ("active")) or ((stele_get_path(stele_context["user"], ["status"])) == ("suspended")) or ((stele_get_path(stele_context["user"], ["status"])) == ("deleted"))


def test_SKU_FORMAT(stele_context):
    result = stele_call_checker("validate-sku", stele_context, {})
    assert result["passed"], result.get("message") or "Checker failed: validate-sku"


def test_EMAIL_FORMAT(stele_context):
    result = stele_call_checker("validate-email", stele_context, {})
    assert result["passed"], result.get("message") or "Checker failed: validate-email"
